modules/phase_field/src/materials/CrossTermBarrierFunctionBase.C

```diff
@@ -0,0 +1,71 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#include "CrossTermBarrierFunctionBase.h"
+
+template<>
+InputParameters validParams<CrossTermBarrierFunctionBase>()
+{
+  InputParameters params = validParams<Material>();
+  params.addParam<std::string>("function_name", "g", "actual name for g(eta_i)");
+  MooseEnum g_order("SIMPLE=0 LOW", "SIMPLE");
+  params.addParam<MooseEnum>("g_order", g_order, "Polynomial order of the barrier function g(eta)");
+  params.addRequiredCoupledVar("etas", "eta_i order parameters, one for each h");
+  params.addRequiredParam<std::vector<Real> >("W_ij", "Terms controlling barrier height set W=1 in DerivativeMultiPhaseMaterial for these to apply");
+  return params;
+}
+
+CrossTermBarrierFunctionBase::CrossTermBarrierFunctionBase(const InputParameters & parameters) :
+    DerivativeMaterialInterface<Material>(parameters),
+    _function_name(getParam<std::string>("function_name")),
+    _g_order(getParam<MooseEnum>("g_order")),
+    _W_ij(getParam<std::vector<Real> >("W_ij")),
+    _num_eta(coupledComponents("etas")),
+    _eta_names(_num_eta),
+    _eta(_num_eta),
+    _prop_g(declareProperty<Real>(_function_name)),
+    _prop_dg(_num_eta),
+    _prop_d2g(_num_eta)
+{
+  // if Vector W_ij is not the correct size to fill the matrix give error
+  if (_num_eta * _num_eta != _W_ij.size())
+    mooseError("Supply the number of etas squared for W_ij.");
```

Comment:
For the error message, how about `Size of W_ij does not match (number of etas)^2. Supply W_ij of correct size.`